pkg/database/resync.go

```diff
@@ -14,65 +16,45 @@ import (
 )
 
 // Reset data for the cluster to the incoming state.
-func (dao *DAO) ResyncData(ctx context.Context, event model.SyncEvent,
-	clusterName string, syncResponse *model.SyncResponse) error {
+func (dao *DAO) ResyncData(ctx context.Context, clusterName string, syncResponse *model.SyncResponse, requestBody []byte) error {
 
-	defer metrics.SlowLog(fmt.Sprintf("Slow resync from %12s.\t RequestId: %d", clusterName, event.RequestId), 0)()
+	defer metrics.SlowLog(fmt.Sprintf("Slow resync from %12s.", clusterName), 0)()
 	klog.Infof(
 		"Starting resync from %12s. This is normal, but it could be a problem if it happens often.", clusterName)
 
 	// Reset resources
-	err := dao.resetResources(ctx, event.AddResources, clusterName, syncResponse)
+	err := dao.resetResources(ctx, clusterName, syncResponse, requestBody)
 	if err != nil {
 		klog.Warningf("Error resyncing resources for cluster %12s. Error: %+v", clusterName, err)
 		return err
 	}
 
 	// Reset edges
-	err = dao.resetEdges(ctx, event.AddEdges, clusterName, syncResponse)
+	err = dao.resetEdges(ctx, clusterName, syncResponse, requestBody)
 	if err != nil {
 		klog.Warningf("Error resyncing edges for cluster %12s. Error: %+v", clusterName, err)
 		return err
 	}
 
-	klog.V(1).Infof("Completed resync of cluster %12s.\t RequestId: %d", clusterName, event.RequestId)
+	klog.V(1).Infof("Completed resync of cluster %12s.", clusterName)
 	return nil
 }
 
 // Reset Resources.
 // 1. Upsert each incoming resource. Keep the UID.
 // 2. Delete existing UIDs that don't match the incoming UIDs.
-func (dao *DAO) resetResources(ctx context.Context, resources []model.Resource, clusterName string,
-	syncResponse *model.SyncResponse) error {
+func (dao *DAO) resetResources(ctx context.Context, clusterName string,
+	syncResponse *model.SyncResponse, resyncBody []byte) error {
 
 	batch := NewBatchWithRetry(ctx, dao, syncResponse)
 
-	incomingUIDs := make([]interface{}, len(resources)+1)
-
 	// UPSERT resources in the database.
-	for i, resource := range resources {
-		uid := resource.UID
-		data, _ := json.Marshal(resource.Properties)
-		query, params, err := useGoqu(
-			"INSERT into search.resources values($1,$2,$3) ON CONFLICT (uid) DO UPDATE SET data=$3 WHERE data!=$3",
-			[]interface{}{uid, clusterName, string(data)})
-		if err == nil {
-			queueErr := batch.Queue(batchItem{
-				action: "addResource",
-				query:  query,
-				uid:    uid,
-				args:   params,
-			})
-			if queueErr != nil {
-				klog.Warningf("Error queuing resources to add. Error: %+v", queueErr)
-				return queueErr
-			}
-			syncResponse.TotalAdded++
-		}
-		incomingUIDs[i] = uid
+	incomingUIDs, err := upsertResources(resyncBody, clusterName, syncResponse, batch)
+	if err != nil {
+		return err
```

Comment:
If there's an error in the UPSERT we won't try to delete.  Maybe this is unlikely to happen, but it could create a situation where records in the DB keep growing.

Reply:
Same happens for edges.